include/zephyr/arch/riscv/arch_inlines.h

```diff
@@ -26,6 +27,17 @@ static ALWAYS_INLINE _cpu_t *arch_curr_cpu(void)
 #endif
 }
 
+#ifdef CONFIG_RISCV_CURRENT_VIA_GP
+register struct k_thread *__arch_current_thread __asm__("gp");
```

Comment:
Style: while asm+register variables are a really useful trick for controlling marshalling behavior around inline assembly, I think here it might be more confusing than valuable and a more traditional asm block which reads the value directly might be clearer.